hazelcast/src/main/java/com/hazelcast/map/impl/query/MapQueryEngineImpl.java

```diff
@@ -219,7 +219,9 @@ protected QueryResult queryParallel(String name, Predicate predicate, Collection
                 = new ArrayList<Future<Collection<QueryableEntry>>>(partitions.size());
 
         for (Integer partitionId : partitions) {
-            QueryPartitionCallable task = new QueryPartitionCallable(name, predicate, partitionId);
+            // Making sure not to use a predicate in a multi-threaded context
+            Predicate predicateCopy = serializationService.copy(predicate);
```

Comment:
am i wrong to assume this copying is only needed at first pass of predicate from user code? `predicate` must have been copied already before coming to `MapQueryEngineImpl `,  for example in `IMap#localKeySet` method so here we are copying _copy of predicate_?